Source/EasyNetQ/Consumer/ConsumerDispatcher.cs

```diff
@@ -8,53 +8,77 @@ namespace EasyNetQ.Consumer
     public class ConsumerDispatcher : IConsumerDispatcher
     {
         private readonly ILog logger = LogProvider.For<ConsumerDispatcher>();
-        private readonly BlockingCollection<Action> queue;
-        private bool disposed;
+        private readonly AutoResetEvent actionAvailable = new AutoResetEvent(false);
+        private readonly ConcurrentQueue<Action> durableActions = new ConcurrentQueue<Action>();
+        private readonly ConcurrentQueue<Action> transientActions = new ConcurrentQueue<Action>();
+        private volatile bool isDisposed;
+
 
         public ConsumerDispatcher(ConnectionConfiguration configuration)
         {
             Preconditions.CheckNotNull(configuration, "configuration");
 
-            queue = new BlockingCollection<Action>();
-
             var thread = new Thread(_ =>
             {
-                while (!disposed && queue.TryTake(out var action, -1))
-                {
+                while (!IsDone())
                     try
                     {
-                        action();
+                        if (durableActions.TryDequeue(out var action) || transientActions.TryDequeue(out action))
+                            action();
+                        else
```

Comment:
What is the purpose of this block? It looks like a spike, to be honest.

Reply:
It should dequeue the action if their is any else wait for a signal :)

Reply:
Why the signal waits for 10000?

Reply:
the 10000 ms is to ensure there is no deadlock :) there should not be any - but better safe than sorry :)

Reply:
@Pliner If you look closely at all cases, then deadlock can really happen. In truth, the likelihood of this is quite small. Therefore, in order not to complicate the code with excessive heavy synchronization constructs, you can do simpler - remove the unconditional wait. It’s simple, it’s clear. And this does not affect performance at all.

Reply:
Could you please describe the scenario of a deadlock?

Reply:
I talked more about https://github.com/EasyNetQ/EasyNetQ/pull/1021/files#r412940189
As for this while loop, the only deadlock situation that I see (theoretical, I said that this is a **very** rare case) is that the thread will be interrupted (aborted) immediately after `isDisposed = true;` so event will be in non-signaled state. Generally speaking, the danger of such loops is that the condition for terminating the loop looks in a completely different variable (`isDisposed` in our case), not in the synchronization primitive (`actionAvailable`). We work with two independent memory areas. Therefore, I always prefer to use timeout for sync primitive so control flow can always return back and check again the loop condition.

I have to admit that in the case of this file, the `Dispose()` method is quite simple and it might seem that I am describing a completely unbelievable situation. But if we assume that more code appears in `Dispose()` later or `Dispose()` becomes virtual, then the chances of deadlock will increase dramatically.

Reply:
I like current solution with cancellation token.